### PR TITLE
Generator for monotonically ordered ulids

### PR DESCRIPTION
**Pull request by camerondavison**
fixes https://github.com/dylanhart/ulid-rs/issues/10

### CODE REVIEW

**Comment by camerondavison** [comment]
added to the bin as well
```
$ cargo run -- -n 10 -m
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
     Running `/Users/cameron/HomeCode/ulid-rs/target/debug/ulid -n 10 -m`
01D28NWKS6TJWHZ748PBENF8Q1
01D28NWKS6TJWHZ748PBENF8Q2
01D28NWKS6TJWHZ748PBENF8Q3
01D28NWKS6TJWHZ748PBENF8Q4
01D28NWKS6TJWHZ748PBENF8Q5
01D28NWKS6TJWHZ748PBENF8Q6
01D28NWKS6TJWHZ748PBENF8Q7
01D28NWKS6TJWHZ748PBENF8Q8
01D28NWKS6TJWHZ748PBENF8Q9
01D28NWKS6TJWHZ748PBENF8QA
```

File: src/lib.rs [hunk 1]
```diff
@@ -199,11 +210,66 @@ impl fmt::Display for Ulid {
...
+    pub fn generate(&mut self) -> Result<Ulid, MonotonicError> {
+        let now = Utc::now();
+        self.generate_from_datetime(now)
+    }
+
+    /// generate a new monotonically ordered ulid. will use previous calls timestamp to
+    /// compare and make sure that each subsequent call is monotonically increasing
+    ///
+    /// # Example
+    /// ```rust
```
Comment (dylanhart) [inline src/lib.rs]: This [doctest](https://doc.rust-lang.org/rustdoc/documentation-tests.html) isn't passing.

File: src/lib.rs [hunk 2]
```diff
@@ -247,6 +313,26 @@ mod tests {
...
+        let ulid3 = Ulid::from_datetime(dt + Duration::milliseconds(1));
+        assert!(ulid1 < ulid2);
+        assert!(ulid2 < ulid3);
+        assert!(ulid2.timestamp_ms() < ulid3.timestamp_ms())
+    }
+
+    #[test]
+    fn test_increment() {
+        let ulid = Ulid::from_string("01BX5ZZKBKZZZZZZZZZZZZZZZX").unwrap();
+        let ulid = ulid.increment().unwrap(); // 01BX5ZZKBKZZZZZZZZZZZZZZZY
```
Comment (dylanhart) [inline src/lib.rs]: Replace the end of line comments with asserts.

File: src/lib.rs [hunk 3]
```diff
@@ -247,6 +313,26 @@ mod tests {
...
     }
 
+    #[test]
+    fn test_order_monotonic() {
+        let dt = Utc::now();
+        let mut gen = Generator::new();
+        let ulid1 = gen.generate_from_datetime(dt).unwrap();
+        let ulid2 = gen.generate_from_datetime(dt).unwrap();
+        let ulid3 = Ulid::from_datetime(dt + Duration::milliseconds(1));
+        assert!(ulid1 < ulid2);
```
Comment (dylanhart) [inline src/lib.rs]: Add `assert_eq!(ulid1.0 + 1, ulid2.0);`

File: src/lib.rs [hunk 4]
```diff
@@ -199,11 +210,66 @@ impl fmt::Display for Ulid {
...
+                    self.previous = Some(next);
+                    return Ok(next.clone())
+                } else {
+                    return Err(MonotonicError::Overflow)
+                }
+            }
+        }
+        let next = Ulid::from_datetime(datetime);
+        self.previous = Some(next);
+        Ok(next.clone())
```
Comment (dylanhart) [inline src/lib.rs]: `Ulid` is copy, so clone is unnecessary.

File: src/lib.rs [hunk 5]
```diff
@@ -199,11 +210,66 @@ impl fmt::Display for Ulid {
...
+    /// ```
+    pub fn generate_from_datetime<T: TimeZone>(&mut self, datetime: DateTime<T>) -> Result<Ulid, MonotonicError> {
+        if let Some(prev) = self.previous {
+            let last_ms = prev.timestamp_ms() as i64;
+            // maybe time went backward, or it is the same ms.
+            // increment instead of generating a new random so that it is monotonic
+            if datetime.timestamp_millis() <= last_ms {
+                if let Some(next) = prev.increment() {
+                    self.previous = Some(next);
+                    return Ok(next.clone())
```
Comment (dylanhart) [inline src/lib.rs]: `Ulid` is copy, so clone is unnecessary.

File: src/lib.rs [hunk 6]
```diff
@@ -144,6 +144,17 @@ impl Ulid {
     pub fn is_nil(&self) -> bool {
         self.0 == 0u128
     }
+
+    /// increment the random number, make sure that the ts millis stays the same
+    fn increment(&self) -> Option<Ulid> {
+        let prev_ts_ms = self.timestamp_ms();
+        let next = Ulid(self.0 + 1);
```
Comment (dylanhart) [inline src/lib.rs]: Addition is not safe here as `self.0` could be `u128::max_value()`. Please mask out the random bits and compare that to the max random bits value.

File: src/lib.rs [hunk 7]
```diff
@@ -247,6 +313,26 @@ mod tests {
...
+        assert!(ulid2 < ulid3);
+        assert!(ulid2.timestamp_ms() < ulid3.timestamp_ms())
+    }
+
+    #[test]
+    fn test_increment() {
+        let ulid = Ulid::from_string("01BX5ZZKBKZZZZZZZZZZZZZZZX").unwrap();
+        let ulid = ulid.increment().unwrap(); // 01BX5ZZKBKZZZZZZZZZZZZZZZY
+        let ulid = ulid.increment().unwrap(); // 01BX5ZZKBKZZZZZZZZZZZZZZZZ
+        assert!(ulid.increment().is_none());
```
Comment (dylanhart) [inline src/lib.rs]: Please add an additional test for overflow:
```
        let ulid = Ulid(u128::max_value());
        let ulid = ulid.increment();
```

Comment (camerondavison) [reply src/lib.rs]: Thanks for being so thorough in your review. I did not know that rust would throw an exception in debug mode for int overflow. the more you know  🎉 :)

File: src/lib.rs (hunk 6 above)
Comment (camerondavison) [reply src/lib.rs]: I changed this it [wrapping_add](https://doc.rust-lang.org/std/primitive.u32.html#method.wrapping_add) if you think that it would be clearer to use the mask and check for max value i can do that instead. `wrapping_add` is what I originally intended.

Comment (kanru) [reply src/lib.rs]: Ulid only has 80 bits of randomness, so you may not want it to overflow to the timestamp bits.

**Comment by camerondavison** [comment]
you are right the mask looks a lot cleaner

**Comment by dylanhart** [comment]
Hi. Sorry I didn't have time to review this yesterday. Going to go through it now.

File: src/lib.rs [hunk 8]
```diff
@@ -51,6 +51,8 @@ pub use base32::EncodingError;
 #[derive(Debug, PartialOrd, PartialEq, Clone, Copy)]
 pub struct Ulid(pub u128);
 
+const MAX_RANDOM: u128 = (1 << 80) - 1;
```
[the remark this answers is not in the document]
Comment (camerondavison) [reply src/lib.rs]: I didn't think about the fact that I could do that. thanks .done

**Comment by dylanhart** [comment]
Thank you!!!
